db/write_batch.cc

```diff
@@ -2613,11 +2605,90 @@ Status WriteBatchInternal::InsertInto(
   return s;
 }
 
+// This class adds checksums for a WriteBatch.
+class ChecksumAdder : public WriteBatch::Handler {
+ public:
+  explicit ChecksumAdder(WriteBatch::ProtectionInfo* prot_info)
+      : prot_info_(prot_info) {}
+
+  ~ChecksumAdder() override {}
+
+  Status PutCF(uint32_t cf, const Slice& key, const Slice& val) override {
+    return add_checksum(cf, key, val, kTypeValue);
+  }
+
+  Status DeleteCF(uint32_t cf, const Slice& key) override {
+    return add_checksum(cf, key, "", kTypeDeletion);
+  }
+
+  Status SingleDeleteCF(uint32_t cf, const Slice& key) override {
+    return add_checksum(cf, key, "", kTypeSingleDeletion);
+  }
+
+  Status DeleteRangeCF(uint32_t cf, const Slice& begin_key,
+                       const Slice& end_key) override {
+    return add_checksum(cf, begin_key, end_key, kTypeRangeDeletion);
+  }
+
+  Status MergeCF(uint32_t cf, const Slice& key, const Slice& val) override {
+    return add_checksum(cf, key, val, kTypeMerge);
+  }
+
+  Status PutBlobIndexCF(uint32_t cf, const Slice& key,
+                        const Slice& val) override {
+    return add_checksum(cf, key, val, kTypeValue);
+  }
+
+  Status MarkBeginPrepare(bool /* unprepare */) override {
+    return Status::OK();
+  }
+
+  Status MarkEndPrepare(const Slice& /* xid */) override {
+    return Status::OK();
+  }
+
+  Status MarkCommit(const Slice& /* xid */) override { return Status::OK(); }
+
+  Status MarkCommitWithTimestamp(const Slice& /* xid */,
+                                 const Slice& /* ts */) override {
+    return Status::OK();
+  }
+
+  Status MarkRollback(const Slice& /* xid */) override { return Status::OK(); }
+
+  Status MarkNoop(bool /* empty_batch */) override { return Status::OK(); }
+
+ private:
+  Status add_checksum(uint32_t cf, const Slice& key, const Slice& val,
+                      const ValueType op_type) {
+    if (prot_info_) {
+      prot_info_->entries_.emplace_back(
+          ProtectionInfo64().ProtectKVO(key, val, op_type).ProtectC(cf));
+    }
+    return Status::OK();
+  }
+
+  // No copy or move.
+  ChecksumAdder(const ChecksumAdder&) = delete;
+  ChecksumAdder(ChecksumAdder&&) = delete;
+  ChecksumAdder& operator=(const ChecksumAdder&) = delete;
+  ChecksumAdder& operator=(ChecksumAdder&&) = delete;
+
+  WriteBatch::ProtectionInfo* const prot_info_ = nullptr;
+};
+
 Status WriteBatchInternal::SetContents(WriteBatch* b, const Slice& contents) {
   assert(contents.size() >= WriteBatchInternal::kHeader);
-  assert(b->prot_info_ == nullptr);
+
   b->rep_.assign(contents.data(), contents.size());
   b->content_flags_.store(ContentFlags::DEFERRED, std::memory_order_relaxed);
+
+  // If we have a prot_info_, compute checksums for the batch.
+  if (b->prot_info_) {
+    ChecksumAdder cs_adder(b->prot_info_.get());
```

Comment:
`RecoverLogFiles()` reuses the same `WriteBatch` object for all the records it decodes from the WAL. As a result, `prot_info_` still has entries for the previous WAL record and then `ChecksumAdder` will append to it. Instead, we need to clear `prot_info_` before using it here.

Reply:
Thanks! I am now creating a new WriteBatch object for every WAL record.